### PR TITLE
UITK Main Toolbar

### PR DESCRIPTION
**Pull request by karl-**
### Purpose of this PR

Rewrite the `ProBuilder Window` toolbar with UITK. Brings a pretty sizable performance boost and removes a lot of hairy IMGUI.

There are some minor differences introduced:

- font size and color is now matched with rest of editor
- spacing is slightly adjusted, matching editor settings
- no more custom scrolling solution, we use the normal grabby bar
- no more custom tooltips
- icon mode buttons scale a little nicer than before

| gui | alloc (kb) | time (ms) |
|-----|-------|-----------|
| imgui | 42.6 | 13.55 |
| uitk | 3.5 | 3.25 |

*IMGUI*
| IMGUI | UITK |
|----|----|
|imgui|uitk|
|imgui|uitk|

*IMGUI*
Screenshot 2023-08-14 at 10 35 00 AM

*UITK*
Screenshot 2023-08-14 at 10 35 55 AM

### Links

**Jira:** PBLD-73

### Comments to Reviewers

Note that in *Icon Mode* the scroll rect is not resizing correctly. I'm following up about it on the UITK channel.

### CODE REVIEW

**Review by gabrielw-us — APPROVED** [review]
Awesome! Big step up.

File: Editor/EditorCore/MainToolbar.cs [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+namespace UnityEditor.ProBuilder.UI
+{
+    class ToolbarMenuItem : VisualElement
+    {
+        public new class UxmlFactory : UxmlFactory<ToolbarMenuItem, UxmlTraits> { }
+        public MenuAction action;
+        public bool iconMode;
+
+        public void RefreshContents()
```
Comment (lopezt-unity) [inline Editor/EditorCore/MainToolbar.cs:15]: I have an error in the editor when switching from text to icon mode (2023.3)
image

Comment (lopezt-unity) [reply Editor/EditorCore/MainToolbar.cs:15]: There is also a lag when all the content is switching from vertical to horizontal mode (and the other direction too)
probuilder-layoutchanged

File: Editor/EditorCore/MenuToolToggle.cs [hunk 2]
```diff
@@ -1,172 +0,0 @@
...
-
-namespace UnityEditor.ProBuilder
-{
-    /// <summary>
-    /// Base class from which any action that is represented in the ProBuilder toolbar inherits.
-    /// </summary>
-    /// <remarks>
-    /// A MenuToolToggle is a special action that creates an <see cref="UnityEditor.EditorTools.EditorTool"/> instance and sets it as the active tool.
-    /// </remarks>
-    public abstract class MenuToolToggle: MenuAction
```
Comment (lopezt-unity) [inline Editor/EditorCore/MenuToolToggle.cs:20]: Can we remove that class as this was public API ?

File: Editor/MenuActions/Editors/NewShapeToggle.cs [hunk 3]
```diff
@@ -11,7 +11,7 @@
 
 namespace UnityEditor.ProBuilder.Actions
 {
-    sealed class NewShapeToggle : MenuToolToggle
+    sealed class NewShapeToggle : MenuAction
```
Comment (lopezt-unity) [inline Editor/MenuActions/Editors/NewShapeToggle.cs:14]: Great improvement on that.
Regarding the MenuToolToggle remark though regarding the fact that this PR removes public API by removing MenuToolToggle class.

**Review by lopezt-unity — COMMENTED** [review]
Great improvement.
Some remarks though regarding certain errors I encountered in the console.
+ Questions regarding the API.

File: Content/UI/ToolbarLight.uss
[the remark this answers is not in the document]
Comment (karl-) [reply Content/UI/ToolbarLight.uss:1]: This is intentional, the light skin uses the default color.

File: Editor/EditorCore/MainToolbar.cs (hunk 1 above)
Comment (karl-) [reply Editor/EditorCore/MainToolbar.cs:15]: The lag is expected, it's rebuilding the visual element tree.

**Comment by lopezt-unity** [comment]
LGTM with recent changes

**Review by lopezt-unity — APPROVED** [review]
Nice :D

**Review by xianenli30 — APPROVED** [review]
I did some spot check but I'll do a more thorough smoke test when we actually do the pb 6 promotion

**Pull request merged**